### ISSUE TITLE
Make a panel height be dependant on the panel's scrollview contentView?

### ISSUE THREAD

**Issue by kevinrenskers** [SCENEE/FloatingPanel#357]
I'm building an app where I need to show a bottom panel like this:

<img width="299" alt="Screen Shot 2020-05-21 at 11 18 37" src="https://user-images.githubusercontent.com/229384/82544126-efbe3680-9b54-11ea-95b4-2ed277ba49e3.png">

This bottom sheet is its own view controller and uses a collectionview since there are a variable number of cells. I am not sure how to make FloatingPanel show this panel "as big as is needed". Using `FloatingPanelIntrinsicLayout` doesn't work, since the collectionview doesn't have an intrinsic height that's set by its cells. So for now I am using a fullscreen layout.

Code is something like this:

``` swift
class SpecialRequestsCoordinator {
  private let floatingPanelController = FloatingPanelController()
  private let rootViewController = SpecialRequestsViewController()

  init() {
    floatingPanelController.delegate = self
    floatingPanelController.isRemovalInteractionEnabled = true
    floatingPanelController.surfaceView.cornerRadius = 15
  }

  func openCatering() {
    let vc = SpecialRequestChoiceSelectViewController()
    floatingPanelController.set(contentViewController: vc)
    rootViewController.present(floatingPanelController, animated: true, completion: nil)
  }
}

extension SpecialRequestsCoordinator: FloatingPanelControllerDelegate {
  func floatingPanel(_ vc: FloatingPanelController, layoutFor newCollection: UITraitCollection) -> FloatingPanelLayout? {
    if vc.contentViewController is SpecialRequestChoiceSelectViewController {
      return FullLayout()
    }

    return IntrinsicLayout()
  }
}

class FullLayout: FloatingPanelLayout {
  func insetFor(position: FloatingPanelPosition) -> CGFloat? {
    switch position {
    case .full: return 0
    default: return nil
    }
  }

  var initialPosition: FloatingPanelPosition {
    return .full
  }

  var supportedPositions: Set<FloatingPanelPosition> {
    return [.full]
  }
}

class IntrinsicLayout: FloatingPanelIntrinsicLayout {
  func insetFor(position: FloatingPanelPosition) -> CGFloat? {
    return nil
  }

  public var initialPosition: FloatingPanelPosition {
    return .full
  }

  public var supportedPositions: Set<FloatingPanelPosition> {
    return [.full]
  }
}
```

And as you might expect, it is now shown like this, which is not ideal. It's way too big and empty:

<img width="395" alt="Screen Shot 2020-05-21 at 11 22 19" src="https://user-images.githubusercontent.com/229384/82544401-5c393580-9b55-11ea-94bd-247f30dd9e84.png">

So how can I configure FloatingPanel to show the panel only as big as a scrollview's `contentSize`? Alternatively I could look into calculating the height in the view controller and pass that into a custom fixed height layout class, but that's going to be a nightmare when taking dynamic type into consideration.

As an aside, it feels like the different `FloatingPanelLayout` protocols (FloatingPanelLayout, FloatingPanelIntrinsicLayout) are quite complicated. If you simply want a fixed height layout, or a 100% fullscreen layout as in my example above, there is so much repeated code.

**Comment by kevinrenskers** [comment]
Okay, got it to work like this:

``` swift
class SpecialRequestChoiceSelectViewController: UIViewController {
  @IBOutlet private var collectionView: UICollectionView!

  func floatingLayoutHeight() -> CGFloat {
    collectionView.layoutIfNeeded()
    return collectionView.collectionViewLayout.collectionViewContentSize.height + 42
  }
}

class FixedHeightLayout: FloatingPanelLayout {
  private let height: CGFloat

  init(height: CGFloat) {
    self.height = height
  }

  func insetFor(position: FloatingPanelPosition) -> CGFloat? {
    switch position {
    case .tip: return height
    default: return nil
    }
  }

  var initialPosition: FloatingPanelPosition {
    return .tip
  }

  var supportedPositions: Set<FloatingPanelPosition> {
    return [.tip]
  }

  func backdropAlphaFor(position: FloatingPanelPosition) -> CGFloat {
    return 0.3
  }
}

extension SpecialRequestsCoordinator: FloatingPanelControllerDelegate {
  func floatingPanel(_ vc: FloatingPanelController, layoutFor newCollection: UITraitCollection) -> FloatingPanelLayout? {
    if let srcsvc = vc.contentViewController as? SpecialRequestChoiceSelectViewController {
      return FixedHeightLayout(height: srcsvc.floatingLayoutHeight())
    }

    return IntrinsicLayout()
  }
}
```

**Comment by ag-hendrix** [comment]
Hey kevinrenskers!

I'm doing something quite similar.  Are you fetching data inside the `SpecialRequestChoiceSelectViewController` to populate the collectionView?  In my case, I'm fetching data from a server but I can't seem to get the height before this is called `func floatingPanel(_ vc: FloatingPanelController, layoutFor newCollection: UITraitCollection)`. Any ideas on how to approach this?

**Comment by kevinrenskers** [comment]
Luckily for me I already have the data before I show the view controller :)

**Comment by FlashTang** [comment]
@kevinrenskers @ag-hendrix 
Why not set constraints for collectionView bottonAnchor to the parent view bottom of the FloatingPanel , it will auto set the collectionView display area , this is my use case code 
```,
NSLayoutConstraint.activate([
            self.tableView.topAnchor.constraint(equalTo: self.view.topAnchor, constant: 34 + CGFloat( TopButtonsView.BAR_HEIGHT)),
            self.tableView.leftAnchor.constraint(equalTo: self.view.leftAnchor, constant: 0),
            self.tableView.bottomAnchor.constraint(equalTo: self.parent!.view!.safeAreaLayoutGuide.bottomAnchor, constant: 0),
            self.tableView.rightAnchor.constraint(equalTo: self.view.rightAnchor, constant: 0)
        ])
```
And make sure  this is inside  did move to parent  , or will crash
```
override func didMove(toParent parent: UIViewController?) {
       //.... active 
}
```
In my case , I didn't track tableview , so you can scroll tableview in both .half and .full mode
Hope it helps

**Comment by ag-hendrix** [comment]
> Hey kevinrenskers!
> 
> I'm doing something quite similar. Are you fetching data inside the `SpecialRequestChoiceSelectViewController` to populate the collectionView? In my case, I'm fetching data from a server but I can't seem to get the height before this is called `func floatingPanel(_ vc: FloatingPanelController, layoutFor newCollection: UITraitCollection)`. Any ideas on how to approach this?

@SCENEE Do you have any input on how I could approach this issue?

**Comment by SCENEE** [comment]
Hello, @kevinrenskers
I'm sorry to my late response. Now I acknowledged this issue.

Hello, @ag-hendrix,

> @SCENEE Do you have any input on how I could approach this issue?

You call `FloatingPanelController.updateLayout()` to call `func floatingPanel(_ vc: FloatingPanelController, layoutFor newCollection: UITraitCollection)`, its method might has some side effect in your content view controller...

I will consider a new configuration or the lib usage for the following use case.
> So how can I configure FloatingPanel to show the panel only as big as a scrollview's contentSize?

**Comment by SCENEE** [comment]
Now I'm planning to add the new API in FloatingPanel v2.0, to adapt the panel height referring to a layout frame, for instance, a scroll view's content size.

**Comment by warpling** [comment]
Amazing!

**Comment by CaramelHeaven** [comment]
@SCENEE 
Has ur idea 'FloatingPanelAdaptiveLayoutAnchor' to 'FloatingPanelLayoutAnchoring' been renamed in v. 2? Maybe I overlooked it c:

**Comment by SCENEE** [comment]
Hi, @CaramelHeaven. `FloatingPanelAdaptiveLayoutAnchor` was released in v2.0.0 and just now I'm implementing it! It will be an object adopting `FloatingPanelLayoutAnchoring` and be released in v2.1.0. Then you will be able to use it with `FloatingPanelLayoutAnchor` and `FloatingPanelIntrinsicLayoutAnchor` :)
